tsconfig.json

```diff
@@ -9,7 +9,8 @@
 		"skipLibCheck": true,
 		"sourceMap": true,
 		"strict": true,
-		"moduleResolution": "Bundler",
+		"module": "NodeNext",
```

Comment:
The SvelteKit team recommends overriding `"module": "NodeNext"` in your tsconfig for svelte library authors.

https://kit.svelte.dev/docs/packaging#caveats

Reply:
I'm only referring to removing `module`, not `moduleResolution`. The recommendation in the docs is to set `"moduleResolution": "NodeNext"`:
> Setting "moduleResolution": "NodeNext" in your tsconfig.json or jsconfig.json will help you with this.

We want `"moduleResolution": "NodeNext"` to help identify malformed import paths, but AFAIK, we _don't_ want `"module": "NodeNext"` as it has the _possibility_ of changing the emitted code from `tsc` to an invalid module format for the browser (e.g. using `require`).

Reply:
Sorry, I should have gone into more detail.  I'm intentionally overriding both`module` and `moduleResolution` to align the library with the TypeScript and SvelteKit best practices for library authors. `module: NodeNext` implies `moduleResolution: NodeNext`, but since bot `module` and `moduleResolution` are set in .svelte-kit/tsconfig.json, we have to override both.

Here's more detail on why these should be set. From the TypeScript handbook:

> When compiling an app, you choose the moduleResolution option for a TypeScript project based on who the module resolution [host](https://www.typescriptlang.org/docs/handbook/modules/theory.html#module-resolution-is-host-defined) is. When compiling a library, you don’t know where the output code will run, but you’d like it to run in as many places as possible. Using "module": "nodenext" (along with the implied ["moduleResolution": "nodenext"](https://www.typescriptlang.org/docs/handbook/modules/reference.html#node16-nodenext)) is the best bet for maximizing the compatibility of the output JavaScript’s module specifiers, since it will force you to comply with Node.js’s stricter rules for import module resolution.
>
>...
>
> In short, "moduleResolution": "bundler" is infectious, allowing code that only works in bundlers to be produced. Likewise, "moduleResolution": "nodenext" is only checking that the output works in Node.js, but in most cases, module code that works in Node.js will work in other runtimes and in bundlers.

https://www.typescriptlang.org/docs/handbook/modules/theory.html#module-resolution-for-libraries

They are a little more direct in the reference section of the TypeScript docs.

> A common misconception is that node16 and nodenext only emit ES modules. In reality, node16 and nodenext describe versions of Node.js that support ES modules, not just projects that use ES modules. Both ESM and CommonJS emit are supported, based on the [detected module format](https://www.typescriptlang.org/docs/handbook/modules/reference.html#module-format-detection) of each file. Because node16 and nodenext are the only module options that reflect the complexities of Node.js’s dual module system, **they are the only correct module options for all apps and libraries that are intended to run in Node.js v12 or later, whether they use ES modules or not.**

https://www.typescriptlang.org/docs/handbook/modules/reference.html#node16-nodenext

It doesn't seem possible to combine `module: esnext` with `moduleResolution: NodeNext`.

> --moduleResolution node16 and nodenext must be paired with their [corresponding module value](https://www.typescriptlang.org/docs/handbook/modules/reference.html#node16-nodenext).

https://www.typescriptlang.org/docs/handbook/modules/reference.html#node16-nodenext-1

The inverse is also true. The bundler option only works with esnext.

> --moduleResolution bundler must be paired with the --module esnext option.

And finally, in the Guide, there is a section titled: "[I'm writing a library](https://www.typescriptlang.org/docs/handbook/modules/guides/choosing-compiler-options.html#im-writing-a-library)", it says to use `module: node16` and explains why this is the best choice for compatibility.

Hopefully this makes sense.

Reply:
Thanks for the clarification. I've gone back to see how the `Library project` template is generated in the latest version of the `create-svelte` CLI and it looks like they've changed it to include `"module": "NodeNext"` now: https://github.com/sveltejs/kit/pull/11029.

Reply:
@rsweeneydev For clarity, what I said before: 
> We want "moduleResolution": "NodeNext" to help identify malformed import paths, but AFAIK, **we don't want "module": "NodeNext" as it has the possibility of changing the emitted code from tsc to an invalid module format for the browser (e.g. using require)**.

used to be the case until [`@sveltejs/package@2.2.3`](https://github.com/sveltejs/kit/pull/11092). Setting `"module": "NodeNext"` _before_ `@sveltejs/package@2.2.3` [would emit CJS](https://github.com/sveltejs/kit/issues/11086), causing things to break in the browser. Looks like they patched it recently. So setting both to `NodeNext` is now OK to do!